### PR TITLE
Enable completion time-to-live to be set on all jobs

### LINKED ISSUE #406: Add a mechanism to clean completed/failed jobs triggered by the jaeger-operator
At this moment there is no mechanism to delete completed/failed jobs(pods) from the namespace.

We can keep the last job [or] provide an option(number of jobs to be retained) in the CR file.

image

### PR DESCRIPTION
**Pull request by objectiser**
Fixes #406 

@jkandasa Could you give this a test to make sure it fixes the issue. Note it [requires kubernetes 1.12 or higher](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#clean-up-finished-jobs-automatically).

Signed-off-by: Gary Brown <gary@brownuk.com>

### CODE REVIEW

**Comment by objectiser** [comment]
Codeclimate issues are related to the autogenerated code.

**Comment by jkandasa** [comment]
@objectiser I do not have 1.12 Kubernetes,

I have, 
```
OpenShift Master: v3.11.98
Kubernetes Master: v1.11.0+d4cacc0
OpenShift Web Console: v3.11.98
```

Test status:
* works well for `es-index-cleaner` job (job status success all the time)
* `spark-dependencies` inconsistent (job status error all the time, #405 )
     * set of old jobs(`spark-dependencies`) are deleted at random interval (could not predict)

CR:
```yaml
  storage:
    type: elasticsearch
    esIndexCleaner:
      enabled: true
      schedule: "*/1 * * * *"
      completedTTL: 180
    dependencies:
      enabled: true
      schedule: "*/1 * * * *"
      completedTTL: 300
    elasticsearch:
      nodeCount: 3
      resources:
```

image

**Comment by objectiser** [comment]
@jkandasa Ok thanks.

@jpkrohling @pavolloffay At present the 'completed time-to-live' for all of the jobs is set to a default of 10 minutes, but wondering whether each job type (e.g. rollover, index cleaner, etc) should have their own defaults which can more closely relate to the schedules (when used) - so if a schedule only runs a job once a day, then maybe the ttl should be two days so any failures are around for a while?

**Comment by pavolloffay** [comment]
Maybe we could make it always double to what is the schedule?

**Comment by objectiser** [comment]
@pavolloffay On further thought, the time the job should remain shouldn't really be related to schedule, as if a failure/retry occurs it can result in a large number of jobs being left lying around for a long time.

On the other hand we need to give time for someone to detect the failure and capture any relevant information. So have updated default to 1 hour.

If we want to try to come up with a more complex scheme, then we could do it in a separate PR?

**Comment by pavolloffay** [comment]
> On the other hand we need to give time for someone to detect the failure and capture any relevant information. So have updated default to 1 hour.

The jobs we have here run once per day at midnight. Not sure if one hour is a good default in this case.

**Comment by objectiser** [comment]
If there is a logging framework, then any failures would be captured centrally - so having the job hanging around would not be as relevant.

**Comment by pavolloffay** [comment]
The question is if those logs would get attention and if they would indicate a problem.

**Comment by objectiser** [comment]
@jpkrohling Any thoughts on this?

**Comment by jpkrohling** [comment]
If we run each job on a daily basis, it's safe to assume that they shouldn't take more than one day to complete.

**Comment by objectiser** [comment]
@jpkrohling Could you take a look at this? codeclimate errors not relevant, and travis job has finished, but not updated here for some reason.

**Comment by jpkrohling** [comment]
This is all green now.

**Comment by objectiser** [comment]
@jpkrohling Are the changes ok to merge?

**Comment by jpkrohling** [comment]
LGTM, but I would prefer `TTLSecondsAfterFinished` as the name for the new property, to match the one from the batch objects, which is where it's used after all.

**Comment by objectiser** [comment]
@jpkrohling No problem, can change.

File: go.sum [hunk 1]
```diff
@@ -43,6 +43,7 @@ github.com/bradfitz/go-smtpd v0.0.0-20170404230938-deb6d6237625/go.mod h1:HYsPBT
 github.com/census-instrumentation/opencensus-proto v0.2.0 h1:LzQXZOgg4CQfE6bFvXGM30YZL1WW/M337pXml+GrcZ4=
 github.com/census-instrumentation/opencensus-proto v0.2.0/go.mod h1:f6KPmirojxKA12rnyqOA5BBL4O983OfeGPqjHWSTneU=
 github.com/client9/misspell v0.3.4/go.mod h1:qj6jICC3Q7zFZvVWo7KLAzC3yx5G7kyvSDkc90ppPyw=
+github.com/codahale/hdrhistogram v0.0.0-20161010025455-3a0bb77429bd h1:qMd81Ts1T2OTKmB4acZcyKaMtRnY5Y44NuXGX2GFJ1w=
```
Comment (objectiser) [inline go.sum]: @jpkrohling Would you expect these changes to the go.sum file?

Comment (jpkrohling) [reply go.sum]: Not really, but the version seems strange. Perhaps it's pointing to master and they had a new commit recently?

Comment (objectiser) [reply go.sum]: Would have expected it to just update the existing line rather than add a new.

Comment (jpkrohling) [reply go.sum]: Oh, good point. I need to understand go modules better in order to answer that. I have seen duplicated entries in different sections in the go.mod, but not sure I would expect the same for go.sum...

**Pull request merged**